controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,293 @@
+			if startOsdCount != expectedOsdCount {
+				storageCluster.Spec.StorageDeviceSets[i].Count = storageCluster.Spec.StorageDeviceSets[i].Count + 1
+			}
+			r.Log.Info("scaling storage cluster", "device set", storageCluster.Spec.StorageDeviceSets[i])
+		}
+	}
+
+	err = r.Patch(ctx, storageCluster, client.MergeFrom(originalStorageCluster))
+	if err != nil {
+		r.Log.Error(err, "failed to update storage cluster")
+		return reconcile.Result{}, err
+	}
+
+	return r.verifyScaling(ctx, storageAutoScaler)
+}
+
+func (r *StorageAutoscalerReconciler) detectLsoStorageclass(ctx context.Context, storageCluster *ocsv1.StorageCluster) error {
+	for _, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		storageclassName := deviceSet.DataPVCTemplate.Spec.StorageClassName
+		// list the storageclass
+		storageClass := &storagev1.StorageClass{}
+		err := r.Get(ctx, client.ObjectKey{Name: *storageclassName}, storageClass)
+		if err != nil {
+			r.Log.Error(err, "failed to get storage class")
+			return err
+		}
+		if storageClass.Provisioner == "kubernetes.io/no-provisioner" {
+			r.Log.Info("storage class has no provisioner")
+			return errs.New("storage class has provisioner as no-provisioner, which is an lso storageclass, autoscaler does not support lso storageclass, delete the autoStorageScaler cr as scaling is not supported")
+		}
+	}
+	return nil
+}
+
+func checkIfScalingRequired(usage interface{}, threshold int) bool {
+	// check if the highest osd usage is greater than or equal to the threshold
+	return ((usage.(float64) * 100) >= float64(threshold))
+}
+
+func calculateExpectedOsdSizeAndCount(storageCluster *ocsv1.StorageCluster, storageAutoScaler *ocsv1.StorageAutoScaler) (resource.Quantity, resource.Quantity, int, int, resource.Quantity, resource.Quantity) {
+	var startOsdSize, expectedOsdSize resource.Quantity
+	var startOsdCount, expectedOsdCount, deviceSetCount, replicaCount int
+
+	// assuming heterogeneous distribution of osd across device sets
+	// so osd size and count is same for all device sets
+	for _, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		if deviceClassMatchesDeviceSet(deviceClass, deviceSet) {
+			startOsdSize = deviceSet.DataPVCTemplate.Spec.Resources.Requests["storage"]
+			replicaCount = deviceSet.Replica
+			startOsdCount = deviceSet.Count * replicaCount
+			deviceSetCount++
+		}
+	}
+
+	expectedOsdSize = startOsdSize
+	expectedOsdCount = startOsdCount
+
+	startStorageCapacity := *resource.NewQuantity(int64(deviceSetCount*startOsdCount)*startOsdSize.Value(), startOsdSize.Format)
+
+	if (resource.NewQuantity(startOsdSize.Value()*2, startOsdSize.Format)).Cmp(storageAutoScaler.Spec.MaxOsdSize) <= 0 {
+		// vertical scaling
+		// double the osd size
+		expectedOsdSize = *resource.NewQuantity(startOsdSize.Value()*2, startOsdSize.Format)
+	} else {
+		// horizontal scaling
+		expectedOsdCount = ((startOsdCount / replicaCount) + 1) * replicaCount
```

Comment:
Is `startOsdCount / replicaCount` guaranteed to be evenly divisible to an integer with no rounding?

```suggestion
		expectedOsdCount = ((startOsdCount / replicaCount) + 1) * replicaCount
```

Reply:
Yes as i am multiplying it before with the same value
`			startOsdCount = deviceSet.Count * replicaCount`